packages/insula/src/ShallowCompare.js

```diff
@@ -0,0 +1,39 @@
+/* eslint complexity: "off" */
+
+function isPrimitive(value) {
+    return Object(value) !== value;
+}
+
+function shallowObj(obj1, obj2) {
+    const obj1Keys = Object.keys(obj1);
+    const obj2Keys = Object.keys(obj2);
+
+    if (obj1Keys.length !== obj2Keys.length) {
+        return false;
+    }
+
+    for (let i = 0; i < obj1Keys.length; i++) {
+        const key = obj1Keys[i];
+        if (obj2.hasOwnProperty(key)) {
+            if (obj1[key] !== obj2[key]) {
+                return false;
+            }
+        } else {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+export function shallowCompare(a, b) {
+    if (typeof a !== typeof b || Array.isArray(a) !== Array.isArray(b)) {
+        return false;
+    }
+
+    if (typeof a === 'function' || isPrimitive(a) || isPrimitive(b)) {
```

Comment:
I think...

```javascript
if (a === b) {
  return true;
} else if (typeof a === 'object' && typeof b === 'object') {
  return shallowObj(a, b);
} else {
  return false;
}
```

which catches primitive equality (first block), object/array shallow compare (second), and functions / anything else (last block). This doesn't account for other classes such as Map or Set, or Immutable's classes, or any other data structure as those `typeof === 'object'` but have no keys and are always "equal". Tempted to change the second `else if` to `else if(isArrayOrObject(a) && isArrayOrObject(b))` where

```javascript
function isArrayOrObject(x) {
  if (isPrimitive(x)) return false;
  return x.constructor === Object || x.constructor === Array;
}
```

Thoughts?

Reply:
That's much cleaner. There are a few edge cases with swapping out empty objects for empty arrays, but I can add a check in the shallowObj func that checks to see the constructors match.